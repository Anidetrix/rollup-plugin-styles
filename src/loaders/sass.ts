import path from "path";
import PQueue from "p-queue";
import { Options as SASSOptions, Result as SASSResult, Importer as SASSImporter, Sass } from "sass";
import { FiberConstructor } from "fibers";

import { Loader, Payload, SASSLoaderOptions } from "../types";
import loadModule from "../utils/load-module";
import resolveAsync from "../utils/resolve-async";

// This queue makes sure node-sass leaves one thread available for executing fs tasks
// See: https://github.com/sass/node-sass/issues/857
const threadPoolSize = parseInt(process.env.UV_THREADPOOL_SIZE || "4");
const workQueue = new PQueue({ concurrency: threadPoolSize - 1 });
const moduleRe = /^~([\da-z]|@).+/i;

type AllowedSassID = "sass" | "node-sass";
const possibleSassIDs: AllowedSassID[] = ["node-sass", "sass"];

const getUrlOfPartial = (url: string): string => {
  const parsedUrl = path.parse(url);
  return `${parsedUrl.dir}${path.sep}_${parsedUrl.base}`;
};

const defaultImporter: SASSImporter = (url, importer, done) => {
  if (!moduleRe.test(url)) return done({ file: url });

  const moduleUrl = url.slice(1);
  const partialUrl = getUrlOfPartial(moduleUrl);

  const options = {
    basedir: path.dirname(importer),
    extensions: [".scss", ".sass", ".css"],
  };

  const finishImport = (id: string): void =>
    // Do not add `.css` extension in order to inline the file
    done({ file: id.replace(/\.css$/i, "") });

  const next = (): void =>
    // Pass responsibility back to other custom importers
    done(null);

  // Give precedence to importing a partial
  resolveAsync(partialUrl, options)
    .then(finishImport)
    .catch(error => {
      if (error.code === "MODULE_NOT_FOUND" || error.code === "ENOENT")
        resolveAsync(moduleUrl, options)
          .then(finishImport)
          .catch(next);
      else next();
    });
};

/**
 * Loads Sass module or throws an error
 * @returns A tuple in format [`loaded sass module`, `id`],
 */
async function loadSassOrThrow(): Promise<[Sass, AllowedSassID]> {
  // Loading one of the supported modules
  for (const id of possibleSassIDs) {
    const module = await loadModule(id);
    if (module) return [module, id];
  }

  // Throwing exception if module can't be loaded
  throw new Error(
    [
      "You need to install one of the following packages:",
      possibleSassIDs.map(id => `\`${id}\``).join(", "),
      "in order to process Sass files",
    ].join("\n"),
  );
}

const loader: Loader<SASSLoaderOptions> = {
  name: "sass",
  test: /\.(sass|scss)$/i,
  async process({ code, map }) {
    const [sass, sassType] = await loadSassOrThrow();

    const render = (options: SASSOptions): Promise<SASSResult> =>
      new Promise((resolve, reject) => {
        sass.render(options, (err, css) => (err ? reject(err) : resolve(css)));
      });

    let fiber: FiberConstructor | undefined;
    // Disable `fibers` for testing, it doesn't work
    if (sassType == "sass" && !process.env.STYLES_TEST) fiber = await loadModule("fibers");

    return workQueue.add<Payload>(
      async (): Promise<Payload> => {
        // node-sass won't produce source maps if the data option is used and `sourceMap` option is not a string.
        // In case it is a string, `sourceMap` option should be a path where the source map is written.
        // But since we're using the data option, the source map will not actually be written, but
        // all paths in sourcemap's sources will be relative to that path.
        const res = await render({
          ...this.options,
          file: this.id,
          data: (this.options.data || "") + code,
          indentedSyntax: /\.sass$/i.test(this.id),
          sourceMap: Boolean(this.sourceMap) && this.id,
          omitSourceMapUrl: true,
          sourceMapContents: true,
          importer: [defaultImporter].concat(this.options.importer || []),
          fiber,
        });

        const deps = res.stats.includedFiles;
        for (const dep of deps) this.dependencies.add(dep);

        return { code: res.css.toString(), map: (res.map && res.map.toString()) || map };
      },
    );
  },
};

export default loader;
