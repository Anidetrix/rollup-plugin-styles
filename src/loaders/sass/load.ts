import loadModule from "../../utils/load-module";
import arrayFmt from "../../utils/array-fmt";

const ids = ["sass", "node-sass"];
const idsFmt = arrayFmt(ids);
export default function (impl?: string): [sass.Sass, string] {
  // Loading provided implementation
  if (impl) {
    const provided = loadModule(impl);
    if (provided) return [provided as sass.Sass, impl];
    throw new Error(`Could not load \`${impl}\` Sass implementation`);
  }

  // Loading one of the supported modules
  for (const id of ids) {
    const sass = loadModule(id);
    if (sass) return [sass as sass.Sass, id];
  }

  throw new Error(`You need to install ${idsFmt} package in order to process Sass files`);
}
