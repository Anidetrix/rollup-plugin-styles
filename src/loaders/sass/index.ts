import PQueue from "p-queue";
import { Options as SASSOptions, Result as SASSResult, Sass } from "sass";
import { FiberConstructor } from "fibers";

import { Loader, Payload, SASSLoaderOptions } from "../../types";
import loadModule from "../../utils/load-module";

import defaultImporter from "./default-importer";

// This queue makes sure node-sass leaves one thread available for executing fs tasks
// See: https://github.com/sass/node-sass/issues/857
const threadPoolSize = parseInt(process.env.UV_THREADPOOL_SIZE || "4");
const workQueue = new PQueue({ concurrency: threadPoolSize - 1 });

type AllowedSassID = "sass" | "node-sass";
const possibleSassIDs: AllowedSassID[] = ["node-sass", "sass"];

/**
 * Loads Sass module or throws an error
 * @returns A tuple in format [`loaded sass module`, `id`],
 */
async function loadSassOrThrow(): Promise<[Sass, AllowedSassID]> {
  // Loading one of the supported modules
  for (const id of possibleSassIDs) {
    const module = await loadModule(id);
    if (module) return [module, id];
  }

  // Throwing exception if module can't be loaded
  throw new Error(
    [
      "You need to install one of the following packages:",
      possibleSassIDs.map(id => `\`${id}\``).join(", "),
      "in order to process Sass files",
    ].join("\n"),
  );
}

const loader: Loader<SASSLoaderOptions> = {
  name: "sass",
  test: /\.(sass|scss)$/i,
  async process({ code, map }) {
    const [sass, sassType] = await loadSassOrThrow();

    const render = (options: SASSOptions): Promise<SASSResult> =>
      new Promise((resolve, reject) => {
        sass.render(options, (err, css) => (err ? reject(err) : resolve(css)));
      });

    let fiber: FiberConstructor | undefined;
    // Disable `fibers` for testing, it doesn't work
    if (sassType == "sass" && !process.env.STYLES_TEST) fiber = await loadModule("fibers");

    return workQueue.add<Payload>(
      async (): Promise<Payload> => {
        // node-sass won't produce source maps if the data option is used and `sourceMap` option is not a string.
        // In case it is a string, `sourceMap` option should be a path where the source map is written.
        // But since we're using the data option, the source map will not actually be written, but
        // all paths in sourcemap's sources will be relative to that path.
        const res = await render({
          ...this.options,
          file: this.id,
          data: (this.options.data || "") + code,
          indentedSyntax: /\.sass$/i.test(this.id),
          sourceMap: Boolean(this.sourceMap) && this.id,
          omitSourceMapUrl: true,
          sourceMapContents: true,
          importer: [defaultImporter].concat(this.options.importer || []),
          fiber,
        });

        const deps = res.stats.includedFiles;
        for (const dep of deps) this.dependencies.add(dep);

        return { code: res.css.toString(), map: (res.map && res.map.toString()) || map };
      },
    );
  },
};

export default loader;
