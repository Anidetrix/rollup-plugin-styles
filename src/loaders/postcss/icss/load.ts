import path from "path";
import fs from "fs-extra";
import { ProcessOptions } from "postcss";
import Processor from "postcss/lib/processor";
import resolveAsync from "../../../utils/resolve-async";

export type Load = (
  url: string,
  file: string,
  extensions: string[],
  processor: Processor,
  opts?: ProcessOptions,
) => Promise<Record<string, string>>;

const load: Load = async (url, file, extensions, processor, opts) => {
  let from: string;
  const options = { basedir: path.dirname(file), extensions };
  try {
    from = await resolveAsync(url, options);
  } catch {
    from = await resolveAsync(`./${url}`, options);
  }

  const source = await fs.readFile(from);
  const { messages } = await processor.process(source, { ...opts, from });

  const exports: Record<string, string> = {};
  for (const msg of messages) {
    if (msg.type !== "icss") continue;
    Object.assign(exports, msg.export as Record<string, string>);
  }

  return exports;
};

export default load;
