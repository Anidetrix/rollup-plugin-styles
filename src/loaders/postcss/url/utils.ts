import { Declaration } from "postcss";
import valueParser, { Node, ParsedValue } from "postcss-value-parser";

const urlFuncRe = /^url$/i;
const imageSetFuncRe = /^(?:-webkit-)?image-set$/i;

export const isDeclWithUrl = (decl: Declaration): boolean =>
  /^(?:url|(?:-webkit-)?image-set)\(/i.test(decl.value);

export const walkUrls = (
  parsed: ParsedValue,
  callback: (url: string, node?: Node) => void,
): void => {
  parsed.walk(node => {
    if (node.type !== "function") return;

    if (urlFuncRe.test(node.value)) {
      const { nodes } = node;
      const [urlNode] = nodes;
      const url = urlNode?.type === "string" ? urlNode.value : valueParser.stringify(nodes);
      callback(url.replace(/^\s+|\s+$/g, ""), urlNode);
      return;
    }

    if (imageSetFuncRe.test(node.value)) {
      for (const nNode of node.nodes) {
        if (nNode.type === "string") {
          callback(nNode.value.replace(/^\s+|\s+$/g, ""), nNode);
          continue;
        }

        if (nNode.type === "function" && urlFuncRe.test(nNode.value)) {
          const { nodes } = nNode;
          const [urlNode] = nodes;
          const url = urlNode?.type === "string" ? urlNode.value : valueParser.stringify(nodes);
          callback(url.replace(/^\s+|\s+$/g, ""), urlNode);
          continue;
        }
      }
    }
  });
};
