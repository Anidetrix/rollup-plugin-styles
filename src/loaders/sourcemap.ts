import { Loader } from "../types";
import { getMap, stripMap } from "../utils/sourcemap";

const loader: Loader<{}> = {
  name: "sourcemap",
  alwaysProcess: true,
  async process({ code, map }) {
    map = (await getMap(code, this.id)) ?? map;
    return { code: stripMap(code), map };
  },
};

export default loader;
