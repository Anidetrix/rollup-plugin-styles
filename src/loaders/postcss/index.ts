import path from "path";
import { RawSourceMap } from "source-map";
import { makeLegalIdentifier } from "@rollup/pluginutils";
import postcss from "postcss";
import loadPostCSSConfig from "postcss-load-config";
import cssnano from "cssnano";

import { Loader, PostCSSLoaderOptions } from "../../types";
import { humanlizePath, normalizePath } from "../../utils/path";
import { mm } from "../../utils/sourcemap";
import resolveAsync from "../../utils/resolve-async";
import safeId from "../../utils/safe-id";
import { denullifyObject, booleanFilter } from "../../utils/filter";

import postcssImport from "./import";
import postcssUrl from "./url";
import postcssModules from "./modules";
import postcssICSS from "./icss";
import postcssNoop from "./noop";

type LoadedConfig = ReturnType<typeof loadPostCSSConfig> extends PromiseLike<infer T> ? T : never;

/**
 * @param id File path
 * @param config `postcss-load-config`'s options
 * @returns Loaded PostCSS config
 */
async function loadConfig(
  id: string,
  config: PostCSSLoaderOptions["config"],
): Promise<Partial<LoadedConfig>> {
  const configPath =
    typeof config === "object" && config.path ? path.resolve(config.path) : path.dirname(id);

  const context: { [prop: string]: object | string } = {
    file: {
      extname: path.extname(id),
      dirname: path.dirname(id),
      basename: path.basename(id),
    },
    options: typeof config === "object" ? config.ctx ?? {} : {},
  };

  return loadPostCSSConfig(context, configPath).catch(error => {
    if (!(error.message as string).toLowerCase().includes("no postcss config found")) throw error;
    return {};
  });
}

const loader: Loader<PostCSSLoaderOptions> = {
  name: "postcss",
  alwaysProcess: true,
  async process({ code, map, extracted }) {
    const { options } = this;

    const config = await loadConfig(this.id, options.config);

    const plugins = [
      ...[
        options.import && postcssImport({ extensions: options.extensions, ...options.import }),
        options.url && postcssUrl(options.url),
      ].filter(booleanFilter),
      ...(options.postcss.plugins ?? []),
      ...(config.plugins ?? []),
    ];

    const autoModules =
      options.autoModules &&
      (typeof options.autoModules === "function"
        ? options.autoModules(this.id)
        : options.autoModules instanceof RegExp
        ? options.autoModules.test(this.id)
        : /\.module\.[A-Za-z]+$/.test(this.id));

    const supportModules = Boolean(options.modules || autoModules);

    const modulesExports: { [filepath: string]: { [prop: string]: string } } = {};

    const postcssOpts: PostCSSLoaderOptions["postcss"] & {
      from: string;
      to: string;
      map: postcss.ProcessOptions["map"];
    } = {
      ...denullifyObject((config.options ?? {}) as Required<postcss.ProcessOptions>),
      ...denullifyObject(options.postcss),
      from: this.id,
      to: options.to ?? this.id,
      map: {
        inline: false,
        annotation: false,
        sourcesContent: true,
        prev: mm(map).relative(path.dirname(this.id)).toObject(),
      },
    };

    delete postcssOpts.plugins;

    if (supportModules) {
      const modulesOptions = typeof options.modules === "object" ? options.modules : {};
      plugins.push(
        ...postcssModules({
          // Skip hash while testing since CSS content would differ on Windows and Linux
          // due to different line endings.
          generateScopedName: process.env.NODE_ENV === "test" ? "[name]_[local]" : undefined,
          failOnWrongOrder: true,
          ...modulesOptions,
        }),
        postcssICSS({
          extensions: options.extensions,
          getReplacements(file, replacements, out) {
            modulesExports[file] = replacements;
            if (
              typeof options.modules === "object" &&
              typeof options.modules.getReplacements === "function"
            ) {
              return options.modules.getReplacements(file, replacements, out);
            }
          },
        }),
      );
    }

    if (options.minimize)
      plugins.push(cssnano(typeof options.minimize === "object" ? options.minimize : {}));

    // Avoid PostCSS warning
    if (plugins.length === 0) plugins.push(postcssNoop);

    const res = await postcss(plugins).process(code, postcssOpts);

    for (const warning of res.warnings())
      this.warn({ name: warning.plugin, message: warning.text });

    const deps = res.messages.filter(msg => msg.type === "dependency");
    for (const dep of deps) this.deps.add(normalizePath(dep.file));

    const assets = res.messages.filter(msg => msg.type === "asset");
    for (const asset of assets) this.assets.set(asset.to, asset.source);

    map = mm((res.map?.toJSON() as unknown) as RawSourceMap)
      .resolve(path.dirname(postcssOpts.to))
      .toString();

    if (!options.extract && this.sourceMap)
      res.css += mm(map)
        .modify(map => void delete map.file)
        .relative()
        .toCommentData();

    if (options.emit) return { code: res.css, map };

    const reservedWords = new Set<string>();
    const saferId = (id: string): string => safeId(id, humanlizePath(this.id));

    const cssExportName = "css";
    reservedWords.add(cssExportName);

    const cssVarName = saferId(cssExportName);
    const modulesVarName = saferId("modules");

    const output = [
      `const ${cssVarName} = ${JSON.stringify(res.css)}`,
      `const ${modulesVarName} = ${JSON.stringify(modulesExports[this.id] ?? {})}`,
      `export const ${cssExportName} = ${cssVarName}`,
      `export default ${supportModules ? modulesVarName : cssVarName}`,
    ];

    if (options.namedExports) {
      const json = modulesExports[this.id];

      const getClassName =
        typeof options.namedExports === "function"
          ? options.namedExports
          : (name: string): string =>
              makeLegalIdentifier(reservedWords.has(name) ? `_${name}` : name);

      for (const name in json) {
        const newName = getClassName(name);

        if (name !== newName)
          this.warn(`Exported \`${name}\` as \`${newName}\` in ${humanlizePath(this.id)}`);

        if (!json[newName]) json[newName] = json[name];

        output.push(`export const ${newName} = ${JSON.stringify(json[name])}`);
      }
    }

    if (options.extract) extracted = { id: this.id, css: res.css, map };

    if (options.inject) {
      if (typeof options.inject === "function") {
        output.push(options.inject(cssVarName, this.id));
      } else {
        const injectorVarName = saferId("injector");
        const injectorId = await resolveAsync("./inject-css", {
          basedir: path.join(
            process.env.NODE_ENV === "test" ? process.cwd() : __dirname,
            "runtime",
          ),
        });
        const injectorData =
          typeof options.inject === "object" ? `,${JSON.stringify(options.inject)}` : "";

        output.push(
          `import ${injectorVarName} from '${normalizePath(injectorId)}'`,
          `${injectorVarName}(${cssVarName}${injectorData})`,
        );
      }
    }

    code = `${output.join(";\n")};`;

    return { code, map, extracted };
  },
};

export default loader;
