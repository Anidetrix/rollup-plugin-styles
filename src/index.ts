import { createFilter } from "@rollup/pluginutils";

import path from "path";
import Concat from "concat-with-sourcemaps";
import { Plugin } from "rollup";
import cssnano, { CssNanoOptions } from "cssnano";
import postcss from "postcss";

import {
  Options,
  PostCSSLoaderOptions,
  LoaderContext,
  Payload,
  ExtractedData,
  ObjectWithUnknownProps,
} from "./types";
import ensurePostCSSOption from "./utils/ensure-postcss-option";
import Loaders from "./loaders";
import { relativePath, normalizePath } from "./utils/path-utils";
import { MapModifier } from "./utils/sourcemap-utils";

/**
 * Infer options from `boolean` or `object`
 * @param option Option
 * @param defaultValue Default value
 * @returns `object` if `option` is truthy, `false` if `option` is `false`, otherwise `defaultValue`
 */
function inferOption<T extends boolean | object>(option: T | undefined, defaultValue: T): T | {} {
  if (typeof option === "boolean") return option === false ? false : {};
  if (typeof option === "object") return option;
  return defaultValue;
}

/**
 * Make sure option is defined
 * @param option Option
 * @param defaultValue Default value
 * @returns `option` if option is defined, or `defaultValue`
 */
function ensureOption<T>(option: T | undefined, defaultValue: T): T {
  return typeof option !== "undefined" ? option : defaultValue;
}

export default (options: Options = {}): Plugin => {
  const filter = createFilter(options.include, options.exclude);
  const postcssPlugins = Array.isArray(options.plugins)
    ? options.plugins.filter(Boolean)
    : options.plugins;

  const postcssLoaderOptions: PostCSSLoaderOptions = {
    inject:
      typeof options.inject === "function" ? options.inject : inferOption(options.inject, true),
    extract: ensureOption(options.extract, false),
    modules: inferOption(options.modules, false),
    namedExports: ensureOption(options.namedExports, false),
    autoModules: ensureOption(options.autoModules, false),
    minimize: inferOption(options.minimize, false),
    config: inferOption(options.config, {}),
    extensions: ensureOption(options.extensions, [".css", ".sss", ".pcss"]),
    postcss: {
      parser: ensurePostCSSOption(options.parser),
      syntax: ensurePostCSSOption(options.syntax),
      stringifier: ensurePostCSSOption(options.stringifier),
      plugins: postcssPlugins,
    },
  };

  let use: (string | [string, ObjectWithUnknownProps])[] = ["sass", "stylus", "less"];

  if (Array.isArray(options.use)) {
    use = options.use;
  } else if (typeof options.use === "object") {
    use = [
      ["sass", options.use.sass || {}],
      ["stylus", options.use.stylus || {}],
      ["less", options.use.less || {}],
    ];
  }

  use.unshift(["postcss", postcssLoaderOptions], "sourcemap");

  const loaders = new Loaders({
    use,
    loaders: options.loaders,
    extensions: postcssLoaderOptions.extensions,
  });

  const extracted = new Map<string, NonNullable<Payload["extracted"]>>();

  const plugin: Plugin = {
    name: "postcss",

    async transform(code, id) {
      if (!filter(id) || !loaders.isSupported(id)) return null;

      if (typeof options.onImport === "function") options.onImport(code, id);

      const loaderContext: LoaderContext = {
        id,
        sourceMap: options.sourceMap,
        dependencies: new Set<string>(),
        warn: this.warn.bind(this),
        error: this.error.bind(this),
        plugin: this,
        options: {},
      };

      const res = await loaders.process(
        { code, map: this.getCombinedSourcemap().toString() },
        loaderContext,
      );

      for (const dep of loaderContext.dependencies) this.addWatchFile(dep);

      if (postcssLoaderOptions.extract) {
        res.extracted && extracted.set(id, res.extracted);
        return {
          code: res.code,
          map: { mappings: "" as "" },
        };
      }

      return {
        code: res.code,
        map: res.map || { mappings: "" as "" },
      };
    },

    augmentChunkHash() {
      if (extracted.size === 0) return;
      return JSON.stringify([...extracted.values()]);
    },

    async generateBundle(opts, bundle) {
      if (extracted.size === 0 || !(opts.dir || opts.file)) return;

      const dir = opts.dir || path.dirname(opts.file || "");
      const file =
        opts.file ||
        path.join(
          opts.dir || "",
          Object.keys(bundle).find(fileName => {
            const record = bundle[fileName];
            return record.type === "chunk" && record.isEntry;
          }) || "",
        );

      const getExtracted = (): ExtractedData => {
        const fileName =
          typeof postcssLoaderOptions.extract === "string"
            ? relativePath(dir, postcssLoaderOptions.extract)
            : `${path.basename(file, path.extname(file))}.css`;
        const fileDir = normalizePath(path.dirname(path.resolve(dir, fileName)));

        const modules = [...this.moduleIds];
        const entries = [...extracted.values()].sort(
          (a, b) => modules.indexOf(a.id) - modules.indexOf(b.id),
        );

        const concat = new Concat(true, path.basename(fileName), "\n");
        for (const res of entries) {
          const relative = relativePath(dir, res.id);
          const map = res.map && new MapModifier(res.map).relative(fileDir).toObject();
          concat.add(relative, res.code, map);
        }

        let code = concat.content.toString();
        const map = concat.sourceMap && new MapModifier(concat.sourceMap);

        if (map) {
          if (options.sourceMap === "inline") {
            code += map.toCommentData();
          } else if (options.sourceMap === true) {
            code += map.toCommentFile(fileName);
          }
        }

        return {
          code,
          map: options.sourceMap === true && map ? map.toString() : undefined,
          codeFileName: fileName,
          mapFileName: `${fileName}.map`,
        };
      };

      if (options.onExtract) {
        const shouldExtract = options.onExtract(getExtracted);
        if (shouldExtract === false) return;
      }

      const res = getExtracted();

      // Perform cssnano on the extracted file
      if (postcssLoaderOptions.minimize) {
        const cssOpts: CssNanoOptions & postcss.ProcessOptions =
          typeof postcssLoaderOptions.minimize === "object" ? postcssLoaderOptions.minimize : {};

        cssOpts.from = res.codeFileName;
        if (options.sourceMap === "inline") {
          cssOpts.map = { inline: true };
        } else if (options.sourceMap === true) {
          cssOpts.map = { prev: res.map };
          cssOpts.to = res.codeFileName;
        }

        const result = await cssnano.process(res.code, cssOpts);
        res.code = result.css;
        if (options.sourceMap === true) res.map = result.map && result.map.toString();
      }

      this.emitFile({
        type: "asset",
        source: res.code,
        fileName: res.codeFileName,
      });

      if (res.map) {
        this.emitFile({
          type: "asset",
          source: res.map,
          fileName: res.mapFileName,
        });
      }
    },
  };

  return plugin;
};
