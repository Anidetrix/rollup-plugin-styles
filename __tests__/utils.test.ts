import { Payload } from "../src/types";
import Loaders from "../src/loaders";
import postcss from "postcss";
import postcssNoop from "../src/loaders/postcss/noop";
import { loadSass } from "../src/loaders/sass/load";
import { ensurePCSSOption } from "../src/utils/options";
import { mm, getMap, stripMap } from "../src/utils/sourcemap";
import { humanlizePath } from "../src/utils/path";

import { fixture } from "./helpers";

jest.mock("../src/utils/load-module", () => jest.fn());
import loadModuleMock from "../src/utils/load-module";

test("noop", async () => {
  const { css } = await postcss(postcssNoop).process(".foo{color:red}", { from: "simple.css" });
  expect(css).toBe(".foo{color:red}");
});

describe("load-module", () => {
  const loadModule = jest.requireActual("../src/utils/load-module")
    .default as typeof loadModuleMock;

  test("wrong path", async () => {
    await expect(loadModule("totallyWRONGPATH/here")).resolves.toBeUndefined();
  });

  test("correct cwd path", async () => {
    await expect(loadModule(humanlizePath(fixture("utils/fixture")))).resolves.toBe(
      "this is fixture",
    );
  });

  test("correct absolute path", async () => {
    await expect(loadModule(fixture("utils/fixture"))).resolves.toBe("this is fixture");
  });

  test("correct path with custom basepath", async () => {
    await expect(loadModule("fixture", fixture("utils"))).resolves.toBe("this is fixture");
  });
});

describe("less", () => {
  test("not found", async () => {
    const loaders = new Loaders({ use: ["less"], loaders: [], extensions: [""] });
    await expect(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      loaders.process({ code: "" }, { id: "file.less" } as any),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe("stylus", () => {
  test("not found", async () => {
    const loaders = new Loaders({ use: ["stylus"], loaders: [], extensions: [""] });
    await expect(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      loaders.process({ code: "" }, { id: "file.styl" } as any),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe("load-sass", () => {
  test("wrong implementation", async () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    await expect(loadSass("swass" as any)).rejects.toThrowErrorMatchingSnapshot();
  });

  test("not found", async () => {
    await expect(loadSass()).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe("loaders", () => {
  test("unlisting", () => {
    const testLoader = {
      name: "test",
      alwaysProcess: true,
      process: (): Payload => ({ code: "" }),
    };
    const loaders = new Loaders({ use: ["test"], loaders: [testLoader], extensions: [""] });
    expect(loaders.getLoader("test")).toBe(testLoader);
    loaders.unlistLoader("test");
    expect(loaders.getLoader("test")).toBeUndefined();
  });
});

describe("option-utils", () => {
  test("wrong postcss option", () => {
    expect(() => ensurePCSSOption("pumpinizer", "plugin")).toThrowErrorMatchingSnapshot();
  });
});

describe("sourcemap-utils", () => {
  test("inline map", async () => {
    let code =
      '.foo {color: red;background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkBAMAAACCzIhnAAAAG1BMVEXMzMyWlpacnJy+vr6jo6PFxcW3t7eqqqqxsbHbm8QuAAAACXBIWXMAAA7EAAAOxAGVKw4bAAAAiklEQVRYhe3QMQ6EIBAF0C+GSInF9mYTs+1ewRsQbmBlayysKefYO2asXbbYxvxHQj6ECQMAEREREf2NQ/fCtp5Zky6vtRMkSJEzhyISynWJnzH6Z8oQlzS7lEc/fLmmQUSvc16OrCPqRl1JePxQYo1ZSWVj9nxrrOb5esw+eXdvzTWfTERERHRXH4tWFZGswQ2yAAAAAElFTkSuQmCC");}';
    await expect(getMap(code)).resolves.toBeUndefined();
    code +=
      "/*# sourceMappingURL=data:application/json;base64,e1RISVM6SVNBU09VUkNFTUFQU0lNVUxBVElPTn0= */";
    await expect(getMap(code)).resolves.toBe("{THIS:ISASOURCEMAPSIMULATION}");
  });

  test("file map", async () => {
    const code = ".foo {color: red;}/*# sourceMappingURL=fixture.css.map */";
    await expect(getMap(code)).rejects.toThrowErrorMatchingSnapshot();
    await expect(getMap(code, "this/is/nonexistant/path.css")).resolves.toBeUndefined();
    await expect(getMap(code, fixture("utils/pointless.css"))).resolves.toBe(
      "{THIS:ISASOURCEMAPSIMULATION}",
    );
  });

  test("strip map", () => {
    const code = ".foo {color: red;}/*# sourceMappingURL=fixture.css.map */";
    expect(stripMap(code)).toBe(".foo {color: red;}");
  });

  test("map modifier", () => {
    const map = JSON.stringify({ sources: ["../a/b/../foo/bar.css", "../b/a/../bar/foo.css"] });
    const relativeSrc = JSON.stringify(mm(map).relative().toObject()?.sources);
    expect(relativeSrc).toBe(JSON.stringify(["../a/foo/bar.css", "../b/bar/foo.css"]));
    expect(mm("thisisnotjson").toString()).toBeUndefined();
  });
});
