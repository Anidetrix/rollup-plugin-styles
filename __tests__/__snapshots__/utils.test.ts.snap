// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`less not found 1`] = `"You need to install \`less\` package in order to process Less files"`;

exports[`load-sass not found 1`] = `"You need to install \`sass\` or \`node-sass\` package in order to process Sass files"`;

exports[`load-sass wrong implementation 1`] = `"Could not load \`swass\` Sass implementation"`;

exports[`option-utils wrong postcss option 1`] = `"Unable to load PostCSS plugin \`pumpinizer\`"`;

exports[`sourcemap-utils file map 1`] = `"Extracted map detected, but no ID is provided"`;

exports[`stylus not found 1`] = `"You need to install \`stylus\` package in order to process Stylus files"`;
